src/main/java/com/aws/greengrass/easysetup/DeviceProvisioningHelper.java

```diff
@@ -390,31 +399,55 @@ public void setupIoTRoleForTes(String roleName, String roleAliasName, String cer
      * @param roleName           name of target role
      * @param rolePolicyName     name of policy to create and attach
      * @param rolePolicyDocument document of policy to create and attach
+     * @param awsRegion          aws region
      * @return ARN of created policy
      */
-    public Optional<String> createAndAttachRolePolicy(String roleName, String rolePolicyName,
-                                                      String rolePolicyDocument) {
-        try {
+    public Optional<String> createAndAttachRolePolicy(String roleName, String rolePolicyName, String rolePolicyDocument,
+                                                      Region awsRegion) {
+        Optional<String> tesRolePolicyArnOptional = getPolicyArn(rolePolicyName, awsRegion);
+        if (tesRolePolicyArnOptional.isPresent()) {
+            outStream.printf("IAM policy named \"%s\" already exists. Please attach it to the IAM role if not "
+                    + "already%n", rolePolicyName);
+            return tesRolePolicyArnOptional;
+        } else {
             String tesRolePolicyArn;
             CreatePolicyResponse createPolicyResponse = iamClient.createPolicy(
                     software.amazon.awssdk.services.iam.model.CreatePolicyRequest.builder().policyName(rolePolicyName)
                             .policyDocument(rolePolicyDocument).build());
             tesRolePolicyArn = createPolicyResponse.policy().arn();
             outStream.printf("IAM role policy for TES \"%s\" created. This policy DOES NOT have S3 access, please "
-                            + "modify it with your private components' artifact buckets/objects as needed when you "
+                    + "modify it with your private components' artifact buckets/objects as needed when you "
                     + "create and deploy private components %n", rolePolicyName);
             outStream.println("Attaching IAM role policy for TES to IAM role for TES...");
             iamClient.attachRolePolicy(
                     AttachRolePolicyRequest.builder().roleName(roleName).policyArn(tesRolePolicyArn).build());
             return Optional.of(tesRolePolicyArn);
-        } catch (EntityAlreadyExistsException e) {
-            // TODO: [P41215965] get and reuse the policy. non trivial because we can only get IAM policy by ARN
-            outStream.printf("IAM policy named \"%s\" already exists. Please attach it to the IAM role if not "
-                    + "already%n", rolePolicyName);
-            return Optional.empty();
         }
     }
 
+    private Optional<String> getPolicyArn(String policyName, Region awsRegion) {
+        String partition = awsRegion.metadata().partition().id();
+        try {
+            String policyArn = String.format(MANAGED_IAM_POLICY_ARN_FORMAT, partition, policyName);
```

Comment:
is it possible to use the `Arn` datatype instead of building it ourselves using string formatting?

Reply:
I made this change in previous revision but had to revert it because it turns out that arns built like that(e.g. `arn:aws:iam:null:null:policy/GreengrassV2TokenExchangeRoleAccess`) do not work well with global services like IAM and when the ARNs are different based on managed vs customer created resources.